test/extended/dr/leader_change.go

```diff
@@ -0,0 +1,113 @@
+package dr
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	promclient "github.com/openshift/origin/test/extended/prometheus/client"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+	apps "k8s.io/kubernetes/test/e2e/upgrades/apps"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/disruption"
+)
+
+var _ = g.Describe("[sig-etcd][Feature:DisasterRecovery][Disruptive]", func() {
+	f := framework.NewDefaultFramework("leader-change")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("leader-change")
+
+	g.It("[Feature:EtcdLeaderChange] Cluster should remain functional through etcd leader change", func() {
+		framework.Logf("Verify SSH is available before restart")
+		masters, _ := clusterNodes(oc)
+		o.Expect(len(masters)).To(o.BeNumerically(">=", 3))
+
+		disruption.Run(f, "etcd Leader Coup d'etat", "leader-change",
+			disruption.TestData{},
+			[]upgrades.Test{
+				&upgrades.ServiceUpgradeTest{},
+				&upgrades.SecretUpgradeTest{},
+				&apps.ReplicaSetUpgradeTest{},
+				&apps.StatefulSetUpgradeTest{},
+				&apps.DeploymentUpgradeTest{},
+				&apps.DaemonSetUpgradeTest{},
+			},
+			func() {
+				prometheus, err := promclient.NewE2EPrometheusRouterClient(oc)
+				o.Expect(err).ToNot(o.HaveOccurred())
+				err = wait.Poll(30*time.Second, 30*time.Minute, func() (done bool, err error) {
+					framework.Logf("Checking for etcdLeader at %v)", time.Now())
+					etcdLeaderPodName, etcdLeaderContainerID, err := getEtcdLeader(oc, prometheus)
+					if err != nil {
+						framework.Logf("getEtcdLeader: error %v)", err)
+						return false, nil
+					}
+					etcdLeaderNodeName := strings.TrimPrefix(etcdLeaderPodName, "etcd-")
+					framework.Logf("Removing etcd leader from pod %q on node %q with containerID %q)", etcdLeaderPodName, etcdLeaderNodeName, etcdLeaderContainerID)
+					for _, node := range masters {
+						framework.Logf("Checking master: error %v)", err)
+						if node.Name == etcdLeaderNodeName {
+							framework.Logf("Removing etcd leader from pod %q on node %q with containerID %q)", etcdLeaderPodName, node.Name, etcdLeaderContainerID)
+							cmd := fmt.Sprintf("sudo -i /bin/bash -cx 'crictl stop %v'", etcdLeaderContainerID)
+							expectSSH(cmd, node)
+							return false, nil
+						}
+					}
+					framework.Logf("leader %s not removed", etcdLeaderPodName)
+					return false, nil
+				})
+				o.Expect(err).NotTo(o.HaveOccurred())
+			},
+		)
+
+	},
+	)
```

Comment:
we can split the test into 2 sections. graceful term leader kill and kill leader from latency.